### ISSUE TITLE
Image cropping with focus points

### ISSUE THREAD

**Issue by thomasaull** [processwire/processwire-requests#150]
I've read in the [official roadmap announcement](http://processwire.com/blog/posts/processwire-2018-roadmap/) there will be support for predefined crops in the core in 2018. I want to opt for a slightly different approach: Instead of predefined crops, I feel like it would make more sense to make use of focus points ([actually there is an outdated module for that](https://github.com/phlppschrr/ImageFocusArea)) 

The reason is simple, in times of responsive websites and images, it doesn't really makes sense to have a predefined dimension for an image. I want different dimensons for mobile, tablet, large screens, … Of course it'd be possible to define multiple crops but the concept of defining a focus point and cropping the images after that, makes much more sense in my opinion. 

<img width="889" alt="my sketches 2 - 2018-01-12 20 01 20" src="https://user-images.githubusercontent.com/227756/34891172-db09be06-f7d4-11e7-805d-90530c81d268.png">

**Comment by gmclelland** [comment]
I agree, I think it could work similar to the https://www.drupal.org/project/focal_point module for Drupal.  Here's an article about it https://www.deeson.co.uk/labs/drupal-focal-point-module-making-most-your-images.

![image](https://user-images.githubusercontent.com/40570/34892872-7aedbc54-f7a0-11e7-892f-a9ec8e2d5b01.png)

**Comment by gmclelland** [comment]
A little more CMS research.

I noticed Umbarco (a .net CMS) is using focal point image cropping.  https://our.umbraco.org/documentation/getting-started/backoffice/property-editors/built-in-property-editors/image-cropper#set-focal-point

With CraftCMS it sounds like https://github.com/smcyr/Craft-FocusPoint is popular:
https://nystudio107.com/blog/creating-optimized-images-in-craft-cms and https://medium.com/parliament-perspective/creating-focal-points-in-craft-cms-f4d4ebd0fb7e

Squarespace.com - https://support.squarespace.com/hc/en-us/articles/205826028-Using-focal-points-to-center-images

wix.com - https://support.wix.com/en/article/changing-the-focal-point-of-an-image-or-video-in-the-wix-pro-gallery#!

Silverstripe - https://www.silverstripe.org/blog/focuspoint-smarter-image-cropping-for-silverstripe/

Sitecore (a .net CMS) - https://inverseproportion.wordpress.com/2017/04/12/image-field-with-focus-cropper-in-sitecore/

I like checking other CMSs because it gives different UI ideas.  I like wix's focal point button
![image](https://user-images.githubusercontent.com/40570/34920590-9b6da034-f93a-11e7-90e4-3585e6decaf8.png)

Hope that helps

**Comment by thomasaull** [comment]
Wow thanks @gmclelland I really hope this idea picks up :)

**Comment by ryancramerdesign** [comment]
I agree, the focal point options seems like a more useful one. Also seems like it would be simpler for editors to define, easier on developers to work with, and more scalable long term—if some dimension or proportion need changed, it wouldn't require defining a new crop, etc. Thanks for the suggestion.

**Comment by gmclelland** [comment]
Haven't tried it, but I came across https://github.com/boschni/FieldtypeFocusPointImage for anyone interested.

**Comment by ryancramerdesign** [comment]
I've added support for this in the latest commit to the dev branch. To use it, click the "Focus" button that appears with every image. You should then get a focus point you can drag around wherever you want it. After saving the page, any `$image->size()` calls that specify *both* width and height will crop to the focus point you defined. In addition, any existing image variations (that specify both width and height, but not crop) will be re-created to reflect the new focus point. 

I think this is pretty useful as it is, but icing on the cake would be the ability to specify a zoom setting as well. For instance, using the drawing in the first post of this thread, adding the ability to not just make the left mountain the center, but also the ability to do those green boxes too (which I'm assuming represent zooms into the photo). So I'm looking at potential ways to implement that as well. 

cc: @horst-n

**Comment by thomasaull** [comment]
Wow! I just tried it, and everything so far works like a charm and is looking good. I'm seriously amazed, how fast this went from request to execution! This basically solves all my needs I did have with images for the last couple of projects :) 

Initially, the different frames in my sketch should illustrate different crops. However, I made a small mistake: the turquoise and green crop could not happen this way – at least one side would have to be on the edge of the original image.

Personally I'm not sure if it does need a zoom function, since you could achieve the same result with a combination of doing a crop first and using the new focus function second. Maybe others have different opinions or I'm missing a more specific use case.

Good work Ryan!

**Comment by jmartsch** [comment]
Hi. I just tried it, and it does not work in my case. Maybe I am doing something wrong.
Set a focal point and saved the page.
![chrome_2018-01-25_12-58-55](https://user-images.githubusercontent.com/551551/35387819-8e526c46-01d1-11e8-857a-c4eab08394fc.png)

Then I embeded the image with
` <img src="<?php echo $newsItem->images->first()->size(480,320)->url; ?>" />` 
and here is what I get:
![chrome_2018-01-25_13-00-40](https://user-images.githubusercontent.com/551551/35387835-a3839f5e-01d1-11e8-949b-f4e79ca8d3ea.png)

Did I do something wrong? Why does the final image not reflect the focal point?

**Comment by ryancramerdesign** [comment]
@thomasaull Thanks, glad you like it, and thanks for the great suggestion. I didn't know about focus point solutions prior to this, but it seems to be a great fit for PW. 

@jmartsch I just tried to duplicate that issue, but can't here. I noticed in your screenshot that you are using something beyond a core images field. I see a download icon next to the filename, and also a "predefined crops" area, neither of which are in the core images field. I'm guessing that's got something to do with it. Can you try switching it to a core images field (or turn off whatever is adding those things to it) and see if the issue remains? Make sure to change the focus point when testing again, as it's not going to re-create image variations unless the focus point changes.

**Comment by ryancramerdesign** [comment]
@jmartsch also double check that it's not simply a browser cache issue. If you've got Chrome dev tools enabled, your browser cache is likely off. Or you can hit reload a couple of times, or use "->URL" uppercase in your API call, which will use a cache busting filename.

**Comment by adrianbj** [comment]
> but icing on the cake would be the ability to specify a zoom setting as well.

Ryan - perhaps you could get some inspiration from this module: https://processwire.com/talk/topic/8079-imagefocusarea/

Maybe you have already seen it, but just in case you haven't.

**Comment by jmartsch** [comment]
It works on a fresh PW installation. So now I have to find the culprit in my existing site. I am using multiple languages, and ImageExtra and CroppableImage3 modules.

**Comment by jmartsch** [comment]
Found the culprit. I had the module ImageFocusArea installed and it caused a conflict. Removed it and everything works fine.

**Comment by horst-n** [comment]
@ryancramerdesign it works nicely in my tests!
I digged not into the code, but want to point out my thoughts early, so please bear with me if something is still answered within the code:

1) when changing the focal point, I sometimes don't get updated variations in the browser output, only when I call $images->removeVariations(). Maybe something like cache-busting URLs would be usefull? (Set to the timestamp of the last modified focal point selection!)

2) I'm definetly for some sort of zoom param. Maybe something basic like additionally to the current behave have a medium and a closeUp setting. (Or one zoom setting with something like a percent value 80 to 20 in steps of 10.)

That would cover everything I currently can think of. :)

**Comment by ryancramerdesign** [comment]
> Personally I'm not sure if it does need a zoom function, since you could achieve the same result with a combination of doing a crop first and using the new focus function second. 

@thomasaull I agree, the focus point is the most important part and the zoom is just icing on the cake. However, if we're going to bake a cake, we might as well put icing on it too... once the cake has cooled. :) So I think I'll work on a zoom solution to add to this a bit later. 

> Ryan - perhaps you could get some inspiration from this module: 

@adrianbj It's a nice module, kind of a hybrid between regular crop and focus point. As far as I can tell, the way it solves the zoom issue is by providing the developer $options argument settings for "align", "inside" and "outside", and previews for the editing user. But in our case, I'm thinking if we have any kind of zoom control, I'd rather it be something the editing user can define on a percentage basis. 

> Found the culprit. I had the module ImageFocusArea installed and it caused a conflict. Removed it and everything works fine.

@jmartsch Great glad that it's working now. Thanks for testing. 

> when changing the focal point, I sometimes don't get updated variations in the browser output, only when I call $images->removeVariations(). Maybe something like cache-busting URLs would be usefull? (Set to the timestamp of the last modified focal point selection!)

@horst-n Are you talking about on the front-end or the back-end? The back-end should already be using cache-busting URLs. But on the front-end, if you wanted a cache busting URL you'd have to use the `$pageimage->URL` property (uppercase URL). 

For the actual thumbnails shown in the admin, we're not using a cropped thumbnail and instead showing the focus point with JS/CSS. This only appears in the square grid mode, since the other modes show the proportional photo. 

> I'm definetly for some sort of zoom param. Maybe something basic like additionally to the current behave have a medium and a closeUp setting. (Or one zoom setting with something like a percent value 80 to 20 in steps of 10.)

@horst-n So I've kind of got zoom working on the JS side. If you want to enable it, edit the InputfieldImage.js file and change [this line](https://github.com/processwire/processwire/blob/dev/wire/modules/Inputfield/InputfieldImage/InputfieldImage.js#L34) from `false` to `true`. Once you do that, you'll get a zoom control that adjusts a box around the focus point, and it'll start saving your zoom setting along with the focus point. The problem is, I haven't been able to figure out how to translate this to something that gets the right result in the ImageSizerEngine modules (though only attempted GD so far). 

Already supported is that `$this->cropping` can be an array with width and height percentages (i.e. `["20%", "50%"]` at index 0 and 1, and they are used for cropping. So I've added that a 3rd element in the array would represent a zoom percentage, i.e. `["20%", "50%", 30]`. The trouble is, I can't figure out how to add support in the `ImageSizerEngine::processResize()`  method. For instance, see my `@todo` and note in ImageSizerEngineGD [here](https://github.com/processwire/processwire/blob/dev/wire/core/ImageSizerEngineGD.php#L255). Any idea how to support this? Thanks.

**Comment by thomasaull** [comment]
@ryancramerdesign I think I found another small feature request: There are cases, where I'm using a background-image with `background-size: contain`, and there it'd make sense to use the focal point as inputs for `background-position`. I don't know if you do it already, but it makes sense for this, to expose the information about the focal point via the API. For me it'd be enough just to use the orignal x/y positions of the source image (and recalculate everything alse by myself) – but maybe there could also be some helpers, which apply it to cropped/resized images automatically

**Comment by ryancramerdesign** [comment]
@thomasaull All Pageimage objects have a `focus()` method that returns an array with "top" and "left" properties, which are percentages (0-100) of top and left, as defined with the image. The default value is 50% for both. These can be used for background-position. Is this what you are looking for?

**Comment by thomasaull** [comment]
@ryancramerdesign Yes! Perfect! I'll give it a shot soon :)

**Comment by gmclelland** [comment]
Great job Ryan!  I like how you dimmed the image to show the focus point better.

I haven't tried CroppableImage3, but I could see somebody later on extending this new focal point feature to show real-time previews of predefined crops based on where the focus point is.  See this video for an example in the Sitecore CMS https://www.youtube.com/watch?v=JxqHYoASUx0&feature=youtu.be  Watch how the thumbnails adjust when the focus point is moved.

**Comment by horst-n** [comment]
@ryancramerdesign the **caching** thing is/was only frontend and with ->URL its solved! :)

For the **zoom** thing, I will have a closer look to it. To your comments in the GD-Engine: the zoom setting needs to be submitted [**here**](https://github.com/processwire/processwire/blob/dev/wire/core/ImageSizerEngineGD.php#L204) as seventh param into [**getResizeMethod**](https://github.com/processwire/processwire/blob/dev/wire/core/ImageSizerEngine.php#L1664) . I will try to have a closer look tomorrow, but the first option can be that it needs to be passed from there into [**getCropDimensions**](https://github.com/processwire/processwire/blob/dev/wire/core/ImageSizerEngine.php#L1388). Only if this doesn't fit there, it needs to be handled after that call as extra calculation, I think. But definetly within the interface getResizeMethod or its dependency calls, and therefore, it is calculated and available in all engines, not only GD. :)

**Comment by ryancramerdesign** [comment]
@horst-n glad the URL thing is solved. :)

Good to hear the logic can go in the getResizeMethod and work with all engines, I didn't realize that. But looking at it now, it makes sense to have that zoom added as an argument there. When you have a chance to take a closer look let me know if there's anything I can do. 

Btw, I'm hoping to get the GIF anim stuff in the next dev version this coming week.

**Comment by horst-n** [comment]
@ryancramerdesign Please can you help me with some informations about the zoom thingy?
With a focal point set on an image, I get $this->options['focus'] set to true, and an array $this->cropping with three entries. 0 and 1 are percent values, like 18.1% and 33.7%. The third value (zoom) is currently 0 for me. What valid values may arrive here, once the zoom feature is enabled?

I will have some small amount of time to go into this today or tomorrow. :)

**Comment by ryancramerdesign** [comment]
@horst-n I think ImageSizerEngines can ignore the "focus" option, as that is one just used by the Pageimage::size() method. It's always going to be true, unless manually set to false from the $options array, which just tells Pageimage::size() to not set the "cropping" option for focus. So I don't think the ImageSizerEngines need to recognize that focus property at all. 

For the "cropping" option, when an array with 0 and 1 as percentages, then the 3rd value is zoom. If 0 or 1, then it means zoom is not used.  If it's 2-100 then it's a zoom percentage (the usable range is probably much smaller I'm guessing, like 5-80 or something). 

If you want to see the zoom populated, the change [this line](https://github.com/processwire/processwire/blob/dev/wire/modules/Inputfield/InputfieldImage/InputfieldImage.js#L34) to true, and make sure your site is in debug mode. Then you'll see a zoom control on images, and you can set it, and I think it should also pass it into the ImageSizerEngine cropping array. Note that the front-end zoom control is not particularly accurate at this stage. Thanks.

**Comment by jmartsch** [comment]
@ryancramerdesign @horst-n @thomasaull I just wanted to say "Thank you" for this 🥇

**Comment by horst-n** [comment]
@ryancramerdesign I have uploaded code for the zoom function for you:

[focus_zoom_additions_for__ImageSizerEngine](http://biriba.de/pw_pop3/focus_zoom_additions_for__ImageSizerEngine.phps)
[focus_zoom_additions_for__ImageSizerEngineGD](http://biriba.de/pw_pop3/focus_zoom_additions_for__ImageSizerEngineGD.phps)

For me it seems to work in all situations, also disabled upscaling is supported, but there may be other or better ways to do it. So feel free to modify it or suggest other possibilities. :)

PS: the zoom value hasn't made it into the imagesizer for me. I had hardcodded it into my tests. So please 
at first check if it get passed into the image sizer. ($this->cropping[2])

**Comment by ryancramerdesign** [comment]
Wow thanks Horst! Just testing it out here and seems to be working well! Nice work.

One case that I can't seem to get working is a square resize (i.e. width=200, height=200) on a portrait image (500x800), like a picture of a skyscraper. Focus settings are: left=50%, top=0%, zoom=60 (though zoom value can be anything that results in visible zoom). Since top=0%, nothing should be cropped out of the top. Yet, the very top of the skyscraper and sky above it still gets cropped out. I think the zoom might need to be affected by the left/top percentages in some way that it isn't right now. Maybe it's just a matter of detecting when top or left is 0%, and not letting the zoom break out of that side? What do you think?

> PS: the zoom value hasn't made it into the imagesizer for me. I had hardcodded it into my tests. So please at first check if it get passed into the image sizer. ($this->cropping[2])

Sorry, there was one other line that needed adjustment. Please delete or comment out this line in wire/core/Pageimage.php. After that it should pass along zoom to to the ImageSizer: 
https://github.com/processwire/processwire/blob/dev/wire/core/Pageimage.php#L597

**Comment by ryancramerdesign** [comment]
@horst-n forgot to tag you in the above message, so adding this just in case you don't get a notification for the message above.

**Comment by horst-n** [comment]
@ryancramerdesign Ryan, I only have expected and tested percent-values. So I can test and implement all possible values later today, or maybe at least tomorrow. That will not be a big deal. When I get automatically passed the zoom value into the image sizer, I can setup a nice testcase for all possibilities with ease. (Not like only one hardcoded value in the code) :)

**Comment by ryancramerdesign** [comment]
@horst-n I was able to fix the issue for the top/left edges by adding the following in the processResize method of ImageSizerEngineGD. 

In the new code you added, `$this->cropping` gets set to true (after cropExtra is populated), so we have to remember the cropping values so we can refer to them later. Before the `if($zoom > 1) {` statement (added by the code you linked earlier), I added this:
~~~~~
$croppingLeft = is_array($this->cropping) ? $this->cropping[0] : 0;
$croppingTop = is_array($this->cropping) ? $this->cropping[1] : 0;
~~~~~
Then further down in the same method, within the `} else if(4 == $resizeMethod) {` statement, I added this:
~~~~~
if($croppingLeft === "0%") $bgX = 0;
if($croppingTop === "0%") $bgY = 0;
~~~~~ 
This probably isn't the best way to do it, but it seemed to solve the issue I mentioned (skyscraper getting cropped out at the top). However, it doesn't solve it for other edges, like right or bottom. I'm guessing the actual fix might be something completely different than this, but I'm not mathematically smart enough to know where/how.

**Comment by horst-n** [comment]
Thanks for the first fix. I will have a look to it later today.

**Comment by horst-n** [comment]
@ryancramerdesign Ryan, I cannot reproduce the issue you said. Maybe it got covered by one little change I made in the [ImageSizerEngineGD](http://biriba.de/pw_pop3/focus_zoom_additions_for__ImageSizerEngineGD.phps) (?):

`$zoom = is_array($this->cropping) && isset($this->cropping[2]) && false !== strpos(strval($this->cropping[0]), '%') && false !== strpos(strval($this->cropping[1]), '%') ? $this->cropping[2] : 0;`

Others than that I haven't changed. Here is the output of an [testimage](http://biriba.de/pw_pop3/pw-zoom-crop-testimage.jpg) via [foreach loops](http://biriba.de/pw_pop3/pw-zoom-crop-foreachloops.phps) for zoom, top, left:

![pw_screen__focus-zoom-cropping](https://user-images.githubusercontent.com/4884546/35944647-6f31d87e-0c5d-11e8-967f-8134863d53bc.jpg)

**Comment by ryancramerdesign** [comment]
@horst-n I also can't duplicate it with your test image above, but I think that's because it's perfectly square. The issue seems to occur on non-square source images. For instance, like the vertical skyscraper photo I mentioned earlier. I'll attach the source image below and screenshots of the result I'm setting.

Source image:
![suntrust11](https://user-images.githubusercontent.com/328272/35969049-9d6e1ad2-0c93-11e8-8ed2-2f692c4e423a.jpg)

Screenshot of my focus/zoom setting: 
<img width="268" alt="screen shot 2018-02-08 at 5 46 43 am" src="https://user-images.githubusercontent.com/328272/35969055-a5e7ffd4-0c93-11e8-84f1-32b1aa80bcf3.png">

Result of a `$image->size(200,200, ['forceNew' => true]);` call. Note how the top of the skyscraper is clipped, as is the sky above it. The crop isn't starting from the top for some reason, despite focus top=0%. 
![suntrust11 200x200](https://user-images.githubusercontent.com/328272/35969140-e50f58b0-0c93-11e8-8e26-6a8c038c68c1.jpg)

**Comment by horst-n** [comment]
Ok, thanks Ryan. I will try and inspect it with that skyscraper image!

FYI: When we "convert" the **cropping**-settings to a **cropExtra**, normally we should set **cropping** to **false**, because after the crop only a resize is necessary. (Cropping is completely done with the intermediate image). But due to possible rounding issues, I have set it **cropping** to **true**, what may result in an optional crop of 1-2 pixel in one direction when downsizing the cropped intermediate to the final image. (This way we can be sure to get the pixel-exact requested result)

ping: @ryancramerdesign

**Comment by horst-n** [comment]
@ryancramerdesign Ok, my calculations for the crop dimensions were totally wrong. :)
Now I have corrected it and it seem to work now for all image ratios. Only a small incorrectness due to rounding issues seem to be in there, (1-3px). But when we finally have decided how to use the zoom function, I can go a bit deeper and solve this to be 100% correct.

### This Post is obsolete! Please go on to the next one. :)

**Comment by horst-n** [comment]
@ryancramerdesign  Ryan, I already has gone ahead and solved the rounding issues. Please ignore my previous code files and use these of version 3:

[focus_zoom_additions_V3_for__ImageSizerEngine](http://biriba.de/pw_pop3/focus_zoom_additions_V3_for__ImageSizerEngine.phps)
[focus_zoom_additions_V3_for__ImageSizerEngineGD](http://biriba.de/pw_pop3/focus_zoom_additions_V3_for__ImageSizerEngineGD.phps)

I have bound the zoom values to be between 2 and 70. Currently you are using 80 in the javascript, maybe you decrease it to 70 there too.
With an image of original dimension = 500 where one want to get a crop diemension = 250 and a zoom greater then 50(!) will result into upscaled (pixelated) variations. With upscaling set to false, the normal behave of PW's imagesizer is to serve a smaller image variation. For the zoom cropping function I think, this will lead to often to undesired outputs, therefor I added the functionality to zoom out and serve a variation with the minimum required dimensions.

I used some images with rulers in it for the testing:
![pw-zoom-crop-testruler-square](https://user-images.githubusercontent.com/4884546/36082241-f775c126-0fa7-11e8-969d-52fa492d768f.jpg)
![pw-zoom-crop-testruler-landscape](https://user-images.githubusercontent.com/4884546/36082244-fce5b24c-0fa7-11e8-81ed-3cdf077c106e.jpg)
![pw-zoom-crop-testruler-portrait](https://user-images.githubusercontent.com/4884546/36082246-000e4c4a-0fa8-11e8-8ee7-518e57b96898.jpg)

And here is a screenshot from the test:
![screen-zoom-crop_ 8](https://user-images.githubusercontent.com/4884546/36082254-114d3818-0fa8-11e8-8ebc-23afcb1766a5.jpg)

**Comment by ryancramerdesign** [comment]
@horst-n Thanks! Just inserted the new code and tested and it seems be be working nicely! Thank you for your work with this. Now I just need to figure out how to get the JS side working as cleanly. :) Do ImageSizerEngineIMagick and/or ImageSizerEngineAnimatedGif need any updates to support this too?

**Comment by horst-n** [comment]
Ryan, the GD and imagick engines need this few code lines integrated:

http://biriba.de/pw_pop3/focus_zoom_additions_V3_for__ImageSizerEngineGD.phps

One question to the zoom value: can we pass a zoom value via API? We can query all values via $image->focus(), but if it is not yet integrated, it would be nice to override the zoom value via API with the image options. 
Yep, the javascript side. Here it would be very nice if the circle center point could be dragged to left 0 and top 0, and not the radius. Also if the zoom value would be displayed as number while dragging the slider may be of help for some cases. 
I really like that new feature allready.

**Comment by ryancramerdesign** [comment]
Thanks @horst-n that was simpler than I realized for IMagick and AnimatedGif, but you are right, I just put the same code in those and tested it out and all works nicely! It is now committed to the dev branch. 

> One question to the zoom value: can we pass a zoom value via API?

Yes, it can be passed in via "cropping" in the $options array to Pageimage::size(). If given an array where index 0 and 1 are percent numbers, and index 2 is the zoom percent, it should be used.Is this what you were asking about, or something different?

> Yep, the javascript side. Here it would be very nice if the circle center point could be dragged to left 0 and top 0, and not the radius.

I see what you mean. I'm not sure if it can be, as I think the draggable is bound by the dimensions of the image, but I will look into it. 

> Also if the zoom value would be displayed as number while dragging the slider may be of help for some cases.

I agree. I was thinking of putting it in somewhere but just hasn't gotten that far with it yet. Will try to get it in there somewhere though.

**Comment by horst-n** [comment]
> Yes, it can be passed in via "cropping" in the $options array to Pageimage::size(). If given an array where index 0 and 1 are percent numbers, and index 2 is the zoom percent, it should be used.Is this what you were asking about, or something different? 

@ryancramerdesign 
I would like to have an option for the zoom only, because when writing API calls, we don't know the left and top values defined  in the image fields. I think what we need is the possibility to override the zoom value from the UI, but without have to define the top and left too. 
The logic in PageImage should be something like: if focus is set and an override value for zoom is passed in, simply override the zoom value. If a zoom value is passed in, but there are no focus settings from the UI, simply ignore it. Does this make sense?

**Comment by ryancramerdesign** [comment]
@horst-n Yes, that makes sense—I will update it to use that logic, and it will detect "zoom" from the options argument.

**Comment by horst-n** [comment]
@ryancramerdesign 
Ryan, I finally solved the UI / JS part too. You can find the files [here](http://biriba.de/pw_pop3/InputfieldImage_v1/?O=D). Please refer to the [notes file](http://biriba.de/pw_pop3/InputfieldImage_v1/InputfieldImage_notes.txt) first! It contains a complete log of all my changes and hacks.
So, please after following and understanding my hacks, (I commented in the code) you have to clean and rearrange it, as my skills in JS are limited. :) Also the part with the square thumbnals grid-view I haven't touched. It behaves different then the zoomBox view part. But the issue with to small thumbnails seems to be solved by my method too. (please check!)

IMHO, the current state looks like my favourite style:
![pw-new-focus-zoom-function_ 01](https://user-images.githubusercontent.com/4884546/36496157-ba515a84-1737-11e8-849a-da983f46e3af.gif)

**Comment by horst-n** [comment]
all the following conversation and colaboration went here: [#commitcomment-27703193](https://github.com/processwire/processwire/commit/8fe1eb13f4cbc85c6d5dad093fc559c438c92667#commitcomment-27703193)
